src/encoded/visualization.py

```diff
@@ -835,3 +983,80 @@ def setZoomLocationLocks(views, view_config, scales_and_center_k):
     # Set the view config.
     view_config["locationLocks"] = locks["location"]
     view_config["zoomLocks"] = locks["zoom"]
+
+def get_chromsize_grid_from_view(view):
+    """ Look through the given view to find a 2d chromsize grid.
+
+    Return the grid information, or return None if it doesn't exist.
+    """
+
+    # Skip if the view lacks a central track.
+    if "center" not in view["tracks"]:
+        return None
+
+    for track in view["tracks"]["center"]:
+        # Skip if the track's center has no contents.
+        if 'contents' not in track:
+            return None
+        if len(track['contents']) == 0:
+            return None
+        # See if any of the contents have the 2d chromosome grid type.
+        chromsize_contents = [ cont for cont in track['contents'] if  cont["type"] == "2d-chromosome-grid"]
+        if len(chromsize_contents) > 0:
+            return chromsize_contents[0]
+    return None
+
+def get_chromsize_grid_from_viewconf(views, files_info):
+    """ Look through the files_info and the views to find the 2d chromsize grid.
+
+    Return the grid information, or return None if it doesn't exist.
+    """
+    # Check all of the views' central tracks and see if any of them are 2d grids.
+
+    for view in views:
+        chromsize_grid = get_chromsize_grid_from_view(view)
+        if chromsize_grid:
+            return chromsize_grid
+
+    # Look through files_info and see if there is a chromsize file.
+    chromsize_files = [ info["item"] for info in files_info if info["item"]["file_format"] == "/file-formats/chromsizes/" ]
+
+    if len(chromsize_files) > 0:
+        # Get the contents for the chromosome grid. We assume there is only 1 chromsize file.
+        chromsize_view, error = create_2d_view(chromsize_files[0])
+        if error:
+            return None
+
+        chromsize_contents = chromsize_view["tracks"]["center"][0]["contents"][0]
+        return chromsize_contents
+
+    return None
+
+def add_2d_chromsize(new_views, files_info):
+    """ If files_info has a chromsize file and new_views contains a 2D view,
+    all of the views will get a 2D chromsize file.
+    """
+
+    chromsize_contents = get_chromsize_grid_from_viewconf(new_views, files_info)
+
+    # If there are no chromsize contents, return
+    if not chromsize_contents:
+        return new_views, None
+
+    # Look through the new_views for any with a central view with contents.
+    views_2d = [ view for view in new_views if len(view["tracks"]["center"][0]["contents"]) > 0 ]
```

Comment:
To fix the issue where chromsizes are added before other center tracks, I used:
```
 views_2d = [view for view in new_views if view["tracks"]["center"] and
                      len(view["tracks"]["center"][0]["contents"]) > 0]
```
This may be getting a bit verbose for list comprehension, but that's a design choice, I guess

Reply:
If I really wanted to I would add an inline function, but it's alright for now. I updated this code and added more descriptive function comments.